in-progress/0003-native-merkle-trees.md

```diff
@@ -0,0 +1,212 @@
+
+Examples of reads are requesting sibling paths, state roots etc.
+
+#### Updates
+
+As a sequencer/prover inserts transaction side-effects, the resulting new state is computed and cached in memory. This allows for the seperation of `committed` and `uncommitted` reads and the easy rolling back of unsuccessful blocks.
+
+#### Commits
+
+When a block settles, the node performs a commit. It verifies any uncommitted state it may have against that published on chain to determine if that state is canonical. If it is not, the `uncommitted` state is dicarded and the node perform an `Update` operation using the newly published side effects.
+
+Once the node has the correct `uncommitted` state, it commits that state to disk. This is the only time that a write transaction is required against the database.
+
+### Updating the World State
+
+The `Update` operation involves inserting side-effects into one or more trees. Depending on the type of tree, we can make significant optimisations to reduce the real-world time taken.
+
+#### Append Only
+
+Append only trees don't support the updating of any leaves. New leaves are inserted at the right-most location and nodes above these are updated to reflect their newly hashed values. Optimisation here is simply a case of dividing the set of leaves into smaller batches and hashing each of these batches into a sub-tree in seperate threads. Finally, the roots are used to build the sub-tree on top before hashing to the root of the main tree.
+
+#### Indexed Tree
+
+Indexed Trees require significantly more hashing than append only trees. In fact, adding a set of leaves to an Indexed Tree finishes with an append only tree insertion of the new leaves. However, before this, it is necessary to update all 'low-value' leaves first.
+
+For each leaf being inserted:
+
+1. Identify the location of the leaf whose value immediately precedes that being inserted.
+2. Retrieve the sibling path of the preceeding leaf before any modification.
+3. Set the 'next' value and index to point to the leaf being inserted.
+4. Set the 'next' value and index of the leaf being inserted to the leaf previously pointed to by the leaf just updated.
+5. Re-hash the updated leaf and update the leaf with this hash, requiring the tree to be re-hashed up to the root.
+
+Unfortunately, this process is very sequential with minimal opportunity for concurrent hashing. Each sibling path must be taken after having updated the 'low leaf' for the previous insertion. We can achieve a reasonable degree of concurrency here though. We first identify all of the 'low-leaf' values that need updating, then for each we merge steps 2 and 5, making a single pass up the tree, retrieving each node's value before overwriting it. We can then schedule each of these tree traversals as a unit of work to be carried out on a thread pool.
```

Comment:
> Each sibling path must be taken after having updated the 'low leaf' for the previous insertion

Whilst this is true for the circuit code, is this true for the native code?
I wonder if there's an optimisation that doesn't update each low-leaf sequentially, but instead figures out which low leaves need to change (based on the new values being inserted), and then computes a strategy for hashing up the tree with the fewest hashes (and reads and writes) of the leaves and their paths as possible? The paths of many low-leaves will share common tree nodes, higher up the tree.

Also, in the case where a single low-leaf would be updated by two new leaf insertions, then we would only want to read and update that leaf and its path once.

I would see the process as:

For each leaf being inserted:

In memory:
 1. Identify the location of the low-leaf whose `next_value` "points over" the `value` being inserted.
 3. Set the `next_value` and `next_index` to point to the leaf being inserted.
 4. Set the `next_value` and `next_index` of the leaf being inserted to the leaf previously pointed to by the leaf just updated.

Then, once we have all the pointers (`next_value` and `next_index`) of all low-leaves and new leaves figured out, retrieve their sibling paths and paths with a batch read (or multiple batch read, if a single batch read isn't big enough), noting that many leaves' sibling paths will share common nodes.

Then do the minimum amount of hashing needed: any tree node that needs to be recomputed (i.e. any node on a path of one of the leaves, including the root) should only be recomputed once.

Reply:
I think there are 2 cases. The first being the general case, where for each low leaf updated, we need to retrieve the sibling path before modifications are made. The second case, when a block is simply being synched can be more optimal as you say. In this case, no sibling paths are required and we can identify common nodes as you say.